### PR TITLE
[Refactor](dialect) Add sql dialect converter plugins

### PR DESCRIPTION
**Pull request by dutyu**
## Proposed changes

The current logic for SQL dialect conversion is all in the `fe-core` module, which may lead to the following issues:
- Changes to the dialect conversion logic may occur frequently, requiring users to upgrade the Doris version frequently within the fe-core module, leading to a longer change cycle.
- The cost of customized development is high, requiring users to replace the fe-core JAR package.

Turning it into a plugin can address the above issues properly.

**Following is the test case:** 
- SparkSQL dialect conterver plugin:

```sql
# install spark-sql converter plugin
MySQL [(none)]> install plugin from "/tmp/sparksql-converter.zip";
Query OK, 0 rows affected (0.14 sec)

MySQL [(none)]> show plugins;
+---------------------------+---------+-------------------------------------------------------------------------+---------+-------------+----------------------------------------------------------------------+--------+---------------------------------------------------------+-----------+------------+
| Name                      | Type    | Description                                                             | Version | JavaVersion | ClassName                                                            | SoName | Sources                                                 | Status    | Properties |
+---------------------------+---------+-------------------------------------------------------------------------+---------+-------------+----------------------------------------------------------------------+--------+---------------------------------------------------------+-----------+------------+
| __builtin_AuditLogBuilder | AUDIT   | builtin audit logger                                                    | 0.12.0  | 1.8.31      | org.apache.doris.qe.AuditLogBuilder                                  | NULL   | Builtin                                                 | INSTALLED | {}         |
| AuditLoader               | AUDIT   | load audit log to olap load, and user can view the statistic of queries | 1.0.0   | 1.8.0       | org.apache.doris.plugin.audit.AuditLoaderPlugin                      | NULL   | /hdpData1/doris/doris_server/fe/plugins/auditloader.zip | INSTALLED | {}         |
| SparkSqlConverter         | DIALECT | SQL dialect converter plugin for spark sql.                             | 1.0.0   | 1.8.0       | org.apache.doris.plugin.dialect.spark.SparkSqlDialectConverterPlugin | NULL   | /tmp/sparksql-converter.zip                             | INSTALLED | {}         |
+---------------------------+---------+-------------------------------------------------------------------------+---------+-------------+----------------------------------------------------------------------+--------+---------------------------------------------------------+-----------+------------+
3 rows in set (0.01 sec)

MySQL [(none)]> show variables like '%dialect%';
+---------------+-----------+---------------+---------+
| Variable_name | Value     | Default_Value | Changed |
+---------------+-----------+---------------+---------+
| sql_dialect   | spark_sql | doris         | 1       |
+---------------+-----------+---------------+---------+
1 row in set (0.02 sec)

# disable enable_fallback_to_original_planner
MySQL [(none)]> set enable_fallback_to_original_planner=false;
Query OK, 0 rows affected (0.02 sec)

# the query succeed, because the sparksql-converter plguin works
MySQL [(none)]> select * from (select 1);
+------+
| 1    |
+------+
|    1 |
+------+
1 row in set (0.03 sec)

# set dialect to doris, so the query failed
MySQL [(none)]> set sql_dialect=doris;
Query OK, 0 rows affected (0.02 sec)

MySQL [(none)]> select * from (select 1);
ERROR 1248 (42000): errCode = 2, detailMessage = Every derived table must have its own alias

# test spark built-in functions
MySQL [(none)]> set sql_dialect=spark_sql;
Query OK, 0 rows affected (0.01 sec)

MySQL [(none)]> select get_json_object('{"a": 1}', '$a');
+--------------------------------+
| json_extract('{"a": 1}', '$a') |
+--------------------------------+
| NULL                           |
+--------------------------------+
1 row in set (0.13 sec)
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/Dialect.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+     */
+    SPARK_SQL("spark_sql"),
+    /**
+     * Hive parser dialect
+     */
+    HIVE("hive"),
+    /**
+     * Iceberg parser dialect
+     */
+    ICEBERG("iceberg"),
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/Dialect.java:49]: iceberg\paimon\hudi\dlf are not query engine, so they don't have sql dialect, remove them

File: fe/fe-core/src/main/java/org/apache/doris/plugin/DialectConverterPlugin.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * Set dialect set which this plugin can handle
+     */
+    ImmutableSet<Dialect> acceptDialects();
+
+    /**
+     * Override this method if you want to convert sql before parse, returns null as default.
+     * */
+    default @Nullable String convertSql(String originSql, SessionVariable sessionVariable) {
+        return null;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/plugin/DialectConverterPlugin.java:43]: how about:
```suggestion
        return originSql;
```

**Comment by dutyu** [comment]
run buildall

**Comment by dutyu** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]

**Review by seawinde — APPROVED** [review]

**Pull request merged**